### PR TITLE
[ADD] hr_work_entry

### PR DESCRIPTION
**Pull request by tivisse**
- hr_work_entry is extracted from hr_payroll to be in its own module

Part of TASK#1904850

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_work_entry/security/ir.model.access.csv [hunk 1]
```diff
@@ -0,0 +1,7 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_hr_work_entry,access_hr_work_entry,model_hr_work_entry,base.group_user,1,1,0,0
+access_hr_work_entry_officer,access_hr_work_entry_officer,model_hr_work_entry,hr.group_hr_user,1,1,1,1
+access_hr_work_entry_type,access_hr_work_entry_type,model_hr_work_entry_type,base.group_user,1,0,0,0
+access_hr_work_entry_type_officer,access_hr_work_entry_type_officer,model_hr_work_entry_type,hr.group_hr_user,1,1,0,0
```
Comment (tivisse) [inline addons/hr_work_entry/security/ir.model.access.csv]: Note: Should we erase those rights with payroll ?

File: addons/hr_work_entry/models/hr_employee.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# -*- coding:utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class HrEmployee(models.Model):
+    _inherit = 'hr.employee'
+    _description = 'Employee'
```
Comment (sswapnesh) [inline addons/hr_work_entry/models/hr_employee.py]: Is really needed ?

File: addons/hr_work_entry/models/hr_work_entry.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    _order = 'display_warning desc,state,date_start'
+
+    name = fields.Char(required=True)
+    active = fields.Boolean(default=True)
+    employee_id = fields.Many2one('hr.employee', required=True, domain=[('contract_ids.state', 'in', ('open', 'pending'))])
+    date_start = fields.Datetime(required=True, string='From')
+    date_stop = fields.Datetime(string='To')
+    duration = fields.Float(compute='_compute_duration', inverse='_inverse_duration', store=True, string="Period")
+    work_entry_type_id = fields.Many2one('hr.work.entry.type')
+    color = fields.Integer(related='work_entry_type_id.color', readonly=True)
```
Comment (sswapnesh) [inline addons/hr_work_entry/models/hr_work_entry.py]:
```suggestion
    color = fields.Integer(related='work_entry_type_id.color')
```

File: addons/hr_work_entry/models/hr_work_entry.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    def _inverse_duration(self):
+        for work_entry in self:
+            if work_entry.date_start and work_entry.duration:
+                work_entry.date_stop = work_entry.date_start + relativedelta(hours=work_entry.duration)
+
+    @api.multi
+    def action_validate(self):
+        work_entries = self.filtered(lambda work_entry: work_entry.state != 'validated')
+        work_entries.write({'display_warning': False})
+        if not work_entries._check_if_error():
```
Comment (sswapnesh) [inline addons/hr_work_entry/models/hr_work_entry.py]: Can't  we use `vals` to call write method once ? (Just asking)

File: addons/hr_work_entry/models/hr_work_entry.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    _name = 'hr.user.work.entry.employee'
+    _description = 'Work Entries Employees'
+
+    user_id = fields.Many2one('res.users', 'Me', required=True, default=lambda self: self.env.user)
+    employee_id = fields.Many2one('hr.employee', 'Employee', required=True)
+    active = fields.Boolean('Active', default=True)
+
+    _sql_constraints = [
+        ('user_id_employee_id_unique', 'UNIQUE(user_id,employee_id)', 'You cannot have twice the same employee.')
```
Comment (sswapnesh) [inline addons/hr_work_entry/models/hr_work_entry.py]:
```suggestion
        ('user_id_employee_id_unique', 'UNIQUE(user_id,employee_id)', 'You cannot have same employee twice.')
```

File: addons/hr_work_entry/models/hr_work_entry.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        for work_entry in self:
+            work_entry.duration = work_entry._get_duration(work_entry.date_start, work_entry.date_stop)
+
+    def _inverse_duration(self):
+        for work_entry in self:
+            if work_entry.date_start and work_entry.duration:
+                work_entry.date_stop = work_entry.date_start + relativedelta(hours=work_entry.duration)
+
+    @api.multi
+    def action_validate(self):
```
Comment (tivisse) [inline addons/hr_work_entry/models/hr_work_entry.py]: The method still exists in enterprise, and has been slightly modified.
It should be
``` python
    @api.multi
    def action_validate(self):
        """
        Try to validate work entries.
        If some errors are found, set `display_warning` to True for conflicting work entries
        and validation fails.
        :return: True if validation succeded
        """
        work_entries = self.filtered(lambda work_entry: work_entry.state != 'validated')
        if not work_entries._check_if_error():
            work_entries.write({'state': 'validated'})
            return True
        return False
```

File: addons/hr_work_entry/models/resource.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# -*- coding:utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import models, fields
+
```
Comment (tivisse) [inline addons/hr_work_entry/models/resource.py]: Missing a second line. If you don't have a linter, I advise you to install one ;)

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        <field name="view_mode">tree,calendar,form,pivot</field>
+    </record>
+
+    <record id="hr_work_entry_action" model="ir.actions.act_window">
+        <field name="name">Work Entry</field>
+        <field name="res_model">hr.work.entry</field>
+        <field name="view_mode">calendar,tree,form,pivot</field>
+    </record>
+
+    <record id="hr_work_entry_calendar" model="ir.ui.view">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]:
```suggestion
    <record id="hr_work_entry_view_calendar" model="ir.ui.view">
```
You can check the coding guidelines for more info.

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                                </t>
+                            </div>
+                        </div>
+                    </t>
+                </templates>
+            </calendar>
+        </field>
+    </record>
+
+    <record id="hr_work_entry_form" model="ir.ui.view">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]: ditto

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                                <field name="duration" nolabel="1" attrs="{'readonly': [('state', '!=', 'draft')]}" /><span class="ml8">Hours</span>
+                            </div>
+                        </group>
+                    </group>
+                </sheet>
+            </form>
+        </field>
+    </record>
+
+    <record id="hr_work_entry_tree" model="ir.ui.view">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]: ditto

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                <field name="work_entry_type_id"/>
+                <field name="duration"/>
+                <field name="state"/>
+                <field name="date_start" string="Beginning"/>
+                <field name="display_warning"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="view_hr_work_entry_filter" model="ir.ui.view">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]: hr_work_entry_view_search

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+                <filter name="date_filter" string="Date" date="date_start"/>
+                <filter name="group_employee" string="Employee" context="{'group_by':'employee_id'}"/>
+                <filter name="group_work_entry_type" string="Type" context="{'group_by':'work_entry_type_id'}"/>
+            </search>
+        </field>
+    </record>
+
+    <!-- HR WORK ENTRY TYPE -->
+
+    <record id="action_view_hr_work_entry_type" model="ir.actions.act_window">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]: hr_work_entry_type_action

File: addons/hr_work_entry/views/hr_work_entry_views.xml [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        <field name="name">Work Entry Types</field>
+        <field name="res_model">hr.work.entry.type</field>
+        <field name="view_mode">kanban,tree,form</field>
+    </record>
+
+    <record id="hr_work_entry_type_tree" model="ir.ui.view">
```
Comment (tivisse) [inline addons/hr_work_entry/views/hr_work_entry_views.xml]: ditto

**Comment by tivisse** [comment]
Another important point, be careful about your commit messages. The structure of the commit message is not correct and the provided pieces of information about the purpose of this are not clear. 

See the guidelines for more information: https://www.odoo.com/documentation/12.0/reference/guidelines.html#commit-message-structure

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo r-

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**